src/Constraint/TypeConstraint.php

```diff
@@ -80,21 +84,15 @@ public function normalize(stdClass $schema, Context $context, Walker $walker)
      */
     public function apply($instance, stdClass $schema, Context $context, Walker $walker)
     {
+        $actualType = gettype($instance);
+
         if (is_string($schema->type)) {
-            if (!Types::isA($instance, $schema->type)) {
+            if (!isset(Types::$phpJsonCompatibility[$actualType.$schema->type])) {
                 $context->addViolation('instance must be of type %s', [$schema->type]);
             }
-        } else {
-            $hasMatch = false;
-
-            foreach ($schema->type as $type) {
-                if (Types::isA($instance, $type)) {
-                    $hasMatch = true;
-                    break;
-                }
-            }
 
-            if (!$hasMatch) {
+        } else {
+            if (!isset(Types::$phpToJson[$actualType]) || !in_array(Types::$phpToJson[$actualType], $schema->type)) {
```

Comment:
Here's an example of problems we could have with this mapping. Suppose you have:

```php
$schema->type = ['string', 'number'];
$actualType = 'integer';
```
In that case, `in_array(Types::$phpToJson[$actualType], $schema->type)` will return false and a violation will be wrongly added to the context.

```